File: src/python/remove_nth_last_node.py
"""
Question
https://www.educative.io/courses/grokking-coding-interview-patterns-python/remove-nth-node-from-end-of-list

Given a singly linked list, remove the nth node from the end of the list and
return its head.

#two-pointers
#slow-fast-pointers
"""


class LinkedListNode:
    def __init__(self, data, next=None):
        self.data = data
        self.next = next


def remove_nth_last_node(head, n: int):
    left = head
    right = head

    for _ in range(n):
        right = right.next

    if not right:
        head = head.next
        return head

    while right.next:
        right = right.next
        left = left.next

    left.next = left.next.next
    return head

File: src/javascript/remove_nth_last_node.js
/*
Question
https://www.educative.io/courses/grokking-coding-interview-patterns-python/remove-nth-node-from-end-of-list

Given a singly linked list, remove the nth node from the end of the list and
return its head.

#two-pointers
#slow-fast-pointers
*/


function remove_nth_last_node(head, n) {
  let left = head;
  let right = head;

  for(let i = 0; i < n; i++) {
    right = right.next;
  }

  if(!right) {
    head = head.next;
    return head;
  }

  while (right.next) {
    left = left.next;
    right = right.next;
  }

  left.next = left.next.next;
  return head;
}
